### PR TITLE
Fix CPU amin/amax backward rematerialization mismatch

### LINKED ISSUE #186799: `torch.compile` silently produces wrong gradient for `atan2(x, y).amin(dim)` on float32
## Summary

`torch.atan2(x, y).amin(dim=-1).sum().backward()` produces a **wrong** `x.grad` under `torch.compile` (Inductor backend). The gradient at one or more argmin positions is dropped (becomes `0.0` or `-0.0`) instead of receiving the correct upstream value.

Forward output matches exactly between eager and compiled. Only the **backward** is wrong, and only in Inductor — `aot_eager` is correct.

Distinct from existing scatter-family stride-0 grad bugs (#179561, #180164, #182012, #183986) — those involve scatter ops feeding into reductions. This bug involves a plain elementwise binary op (`atan2`) feeding into an `amin`/`amax` reduction, no scatter required.

## Minimal reproducer

```python
import torch

torch.manual_seed(0)
x_init = torch.randn(2, 3)
y = torch.randn(2, 3)

def fn(x, y):
    return torch.atan2(x, y).amin(dim=-1)

x_e = x_init.detach().clone().requires_grad_(True)
fn(x_e, y).sum().backward()

x_c = x_init.detach().clone().requires_grad_(True)
torch.compile(fn)(x_c, y).sum().backward()

print("eager grad   :", x_e.grad.flatten().tolist())
print("compiled grad:", x_c.grad.flatten().tolist())
print("max diff:", (x_e.grad - x_c.grad).abs().max().item())
```

Output on `2.13.0.dev20260608+cpu`:

```
eager grad   : [0.0, 0.0, -0.13662527501583099, -0.0, -0.38940736651420593, 0.0]
compiled grad: [0.0, 0.0, -0.0,                 -0.0, -0.38940736651420593, 0.0]
max diff: 0.13662527501583099
                ^^^^^^^^^^^^^^^^^
                argmin position of row 0 — eager has -0.1366, compiled has -0.0
```

Forward output is identical between eager and compiled (no fwd diff).

## Backend isolation

```
aot_eager   : x.grad diff = 0.000000e+00   OK
inductor    : x.grad diff = 1.366253e-01   BUG
```

Bug is exclusive to Inductor.

## Robustness (random sweep)

| Condition           | Result |
|---------------------|--------|
| 100 random (seed, shape) combinations on float32 | **31 / 100 reproduce** the divergence |
| float32             | **BUG** (consistent 30%+ hit rate) |
| float64             | **OK**  (diff=0.0) |
| float16 / bfloat16  | **NaN** in compiled gradient (likely a related lowering bug) |
| dim=0 reduction     | OK |
| dim=-1 reduction (= dim=1 for 2D) | BUG |
| 1D tensors          | OK |

## What does and doesn't trigger the bug

| Forward chain | Backward Bug? |
|---|---|
| `atan2(x, y).amin(dim=-1)` | **BUG** |
| `atan2(x, y).amax(dim=-1)` | **BUG** |
| `atan2(x, y).abs().amin(dim=-1)` | **BUG** |
| `atan2(x, y).square().amin(dim=-1)` | **BUG** |
| `atan2(x, y).min(dim=-1)[0]` | OK (tuple-returning `min`!) |
| `(x - y).abs().amin(dim=-1)` | OK |
| `(x + y).amin(dim=-1)` | OK |
| `atan2(x, y).relu().amin(dim=-1)` | OK |
| `atan2(x, y).sigmoid().amin(dim=-1)` | OK |
| `atan2(x, y).sum()` | OK (any non-`amin`/`amax` reduction) |

Two distinguishing details:

1. **`.min(dim=-1)[0]` is OK, `.amin(dim=-1)` is BUG.** Same semantics, different lowering path in Inductor. The `amin`/`amax` lowering must be missing something the `min`/`max` lowering has.
2. **`atan2` is required.** Replacing it with `add`, `sub`, `mul`, or `div` makes the bug disappear. Other binary primitives also need their derivative to flow through the reduction — `atan2`'s derivative `∂atan2(a,b)/∂a = b/(a²+b²)` is the specific shape Inductor is mishandling.

## Root cause hypothesis

Looking at the output: at the row-0 argmin position, Inductor produces `-0.0` (signed zero) instead of the correct finite value `-0.1366`. This is the signature of a fused `argmin + scatter` codegen path that writes the chain-rule contribution to the wrong location, or writes zero where it should write the derivative.

The fact that `.min(dim)[0]` works but `.amin(dim)` doesn't strongly suggests this is a regression in Inductor's `amin`/`amax` backward decomposition — likely a specialized lowering of `amin` that doesn't correctly recompose with non-trivial elementwise predecessors like `atan2`.

## Versions

```
$ python -c "import torch; print(torch.__version__)"
2.13.0.dev20260608+cpu
```

Also reproduces on `2.11.0+cpu`. 

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186953

Fix CPU amin/amax backward rematerialization mismatch

CPU Inductor can compute value-selecting reductions such as amin and amax through one floating-point math path in the forward reduction while AOTAutograd rematerializes the reduced input in backward through another path. The amin/amax backward formula routes gradients with an exact equality mask against the reduced input. On CPU, scalar and vector libm paths can differ by a few bits, so that exact mask can miss the selected element and silently drop gradients, or divide by a zero match count and produce NaNs.

Materialize the CPU floating reduction input in the AOTAutograd partitioner when an amin/amax input is consumed by the associated backward exact equality mask. The pass follows view-only chains, inserts a clone immediately before the reduction, routes both the forward reduction and the backward mask through that clone, saves the clone across the cut, and forces the mask itself to be recomputed from the saved clone. This intentionally saves the clone even across activation checkpoint boundaries, because correctness depends on the forward reduction and backward mask using the same materialized buffer. Reductions that share an input also share one saved clone.

Leave NodeInfo untouched when no materialization is needed. After inserting a clone, rebuild fw_order with the same dense forward-only numbering used by classify_nodes so unrelated backward or unclaimed nodes cannot distort the min-cut distance heuristics.

This keeps ordinary elementwise vectorized codegen intact. The rejected alternative was scalarizing CPU vector math such as atan2, which fixed the mismatch but regressed the standalone compiled atan2 hot path by about 19x.

Fixes #186799
Generated by my agent

Benchmark Results:
Compiled CPU `torch.atan2(x, y)` on 1024x1024 float32, one thread, using `torch.utils.benchmark.Timer(...).blocked_autorange(min_run_time=2.0)`:
- Previous broad scalarized-codegen attempt: old vector lowering 1.423 ms, scalarized lowering 27.424 ms, 19.27x slower.
- Final partitioner fix: 1.43 ms median, preserving the vectorized atan2 hot path.

Test Plan:
- `ATEN_CPU_CAPABILITY=avx2 TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 python - <<'PY' ... runpy.run_path('test/inductor/test_cpu_repro.py', run_name='__main__') ... PY` with an in-memory torchvision stub for `CPUReproTests.test_cpu_floating_amin_amax_backward_matches_eager`: passed, `Ran 1 test in 19.181s`, `OK`.
- `python - <<'PY' ... runpy.run_path('test/functorch/test_aotdispatch.py', run_name='__main__') ... PY` with the same stub for `TestPartitioning.test_materialize_cpu_floating_aminmax_eq_input_fw_order`: passed, `Ran 1 test in 0.408s`, `OK`.
- Standalone AVX2 CPU Inductor check with one `sin` result feeding both `amin` and `amax`: compiled output and gradients matched eager.
- `TestPartitioning.test_recompute_partitioning` and `TestAOTAutograd.test_default_partitioner_saves_symints_not_tensors_for_bw`: passed.
- Prior cpp-wrapper coverage of the CPU regression under AVX2: passed, `Ran 1 test in 37.092s`, `OK`.
- Prior performance check: compiled CPU `torch.atan2(x, y)` median `1.43 ms`, preserving vectorized performance.
- `lintrunner -a`: passed, `ok No lint issues.`
- `git diff HEAD~ --check` and `git diff --cached --check`: passed.

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I fixed the CI failure in test/inductor/test_cpu_repro.py by changing the new regression test to compare CPU Inductor directly against eager float32 on the same platform, with equal_nan=True, instead of overconstraining platform/libm behavior against an fp64 reference. I also added the requested partitioner comment explaining why saving the producer is sufficient. The TorchTitan failures were dependency download errors for spmd-types and appear unrelated to this PR.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I fixed the remaining CI regression in test_cpu_floating_amin_amax_backward_matches_eager. The partitioner now materializes the CPU floating amin/amax input and recomputes the backward equality mask from that saved clone, covering AVX2 cpp-wrapper, activation-checkpointed producers, and expanded-view inputs while preserving the vectorized atan2 hot path.

**Comment by jansel** [comment]
My agent says I addressed the latest review feedback by documenting the intentional activation-checkpoint save, clarifying that the equality-mask recompute tag is for the min-cut/Inductor path, and renaming the helper so it no longer implies the producer itself is cloned. The follow-up passed the targeted CPU Inductor regression test and lintrunner -a.

File: torch/_functorch/partitioners.py [hunk 1]
```diff
@@ -2319,6 +2323,149 @@ def force_save_bw_mutation_src(joint_module: fx.GraphModule) -> None:
...
+            )
+        ]
+
+    def replace_node_arg(container: Any, old: fx.Node, new: fx.Node) -> Any:
+        return pytree.tree_map(lambda arg: new if arg is old else arg, container)
+
+    def update_fw_order() -> None:
+        if "required_fw_nodes" in node_info.__dict__:
+            del node_info.__dict__["required_fw_nodes"]
+        node_info.fw_order = {
```
Comment (bobrenjc93) [inline torch/_functorch/partitioners.py]: `update_fw_order` regresses `fw_order` semantics for *every* joint graph, not just ones this pass touches. `classify_nodes` builds `fw_order` with a dense forward-only counter (`fw_cnt`), but here `idx` enumerates all graph nodes and only fw nodes are kept, so the values are global node indices. `solve_min_cut` compares those values with absolute thresholds — `node_info.get_fw_order(cur) > start_order + 100` in `ban_if_long_fusible_chains`, and `max_range` in `find_first_unfusible` — so any graph with non-fw nodes interleaved among/before the forward ops (tangent placeholders, `must_be_in_backward`-tagged nodes under FSDP/AC, unclaimed nodes) gets inflated distances and different recomputation bans, on CUDA too, with no amin/amax anywhere. Please (a) use a dense counter matching `classify_nodes`, and (b) only rebuild when the pass actually rewrote something (early-return if no clone was inserted) so the common path is a true no-op.

Comment (jansel) [reply torch/_functorch/partitioners.py]: My agent says fixed in 7bacd8dcad4e: fw_order is now rebuilt with the same dense forward-only counter as classify_nodes, and the pass returns without touching NodeInfo when it inserts no clone. I added focused regression coverage for both paths; it also verifies that shared amin/amax inputs reuse one saved clone.
